File: backend/src/Maran.Agent.Client/Services/AccountsService/QuotaUnenforceableReason.cs
namespace Maran.Agent.Client.Services.AccountsService;

/// <summary>
/// The reason a filesystem cannot hold an enforceable disk quota, as far as the agent can tell
/// from where it is standing. Present only when <see cref="AccountUsageDto.QuotaState"/> is
/// <see cref="AccountQuotaState.NotEnforceable"/>. Mirrors the agent's own
/// <c>ops::accounts::QuotaUnenforceableReason</c>.
/// </summary>
/// <remarks>
/// A network filesystem mounted for <c>/home</c> — NFS, most concretely — can report a LOCAL
/// state disconnected from server-side enforcement (<c>rpc.rquotad</c>); the agent has no rpc path
/// to a remote NFS server's quota daemon, and this is a known gap, not one either side pretends to
/// close.
/// </remarks>
public enum QuotaUnenforceableReason
{
    /// <summary>The agent predates this field, or the state is not <see cref="AccountQuotaState.NotEnforceable"/>.</summary>
    Unspecified = 0,

    /// <summary>
    /// The filesystem is not mounted with a quota-accounting option
    /// (<c>usrquota</c>/<c>uquota</c>/<c>usrjquota</c>).
    /// </summary>
    MountedWithoutQuotaAccounting = 1,

    /// <summary>Mounted with the option, but <c>quotaon -p</c> reports accounting is off.</summary>
    AccountingNotEnabled = 2,
}

File: agent/crates/ops/src/accounts/model/quota_unenforceable_reason.rs
//! What the panel can honestly say about one account's disk quota.

/// The reason a filesystem cannot hold an enforceable disk quota, as far as
/// this agent can tell from where it is standing.
///
/// `#[non_exhaustive]`: a future observation (an NFS-aware check, a third quota
/// mechanism) is added as a new variant, and every `match` in this crate fails
/// to compile until it is handled — the same protection a named catch-all
/// would give, without inventing a variant that means nothing
/// (`agent/crates/ops/src/db/model/grant_repair_refusal.rs`'s idiom).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[non_exhaustive]
pub enum QuotaUnenforceableReason {
    /// The filesystem holding the account's home is not mounted with a quota
    /// accounting option (`usrquota`/`uquota`/`usrjquota`), so no per-user
    /// limit can exist on it no matter what `setquota` is asked to do.
    MountedWithoutQuotaAccounting,

    /// The filesystem is mounted with a quota accounting option, but
    /// `quotaon -p` reports user quota accounting is OFF for it — mounted
    /// correctly but never turned on (or turned off since).
    AccountingNotEnabled,
}
